nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp

```diff
@@ -259,6 +295,61 @@ void BinaryFilter::changeState(const bool state)
     std::make_unique<std_msgs::msg::Bool>();
   msg->data = state;
   binary_state_pub_->publish(std::move(msg));
+  changeParameters(state);
+}
+
+void BinaryFilter::changeParameters(const bool state)
+{
+  for (size_t param_index = 0; param_index < binary_parameters_info_.size(); ++param_index) {
+    if (!change_parameters_clients_.at(param_index)->wait_for_service(
```

Comment:
Also related to previous comment:

> Perform wait for service here to make sure up before continuing

Do you mean that if any service goes timeout at the beginning we should not add it or just give an error/warning? What if the server is temporary offline (e.g. from a plugin)? 
Also, a node could go offline at a certain point and since the service is checked only at initialization time, there would not be information for the user to understand that parameters are not updated.
I understand that adding the `wait_for_server` slows down the execution, but in my opinion it should be the user to keep attention to the selected parameters (e.g. too many "offline" servers).

Reply:
Once the BF is being configured, we need to check that services/nodes are available on that stage. Later on we are assuming that everything is OK, and we are working with bringed-up nodes. If something goes wrong (e.g. node became offline), service request should fail by timeout. For that, see Steve comment from above and please use `spin_until_future_complete(future_result, change_parameter_timeout_)` API to not have infinite loops. Anyway, loosing the service with the node - is extraordinary situation, which could be handled by other than each time checking for service availability way (performance for usual serivce calls in this way will be faster, and we still could handle error situations).

Reply:
Hi there. I agree with using `spin_until_future_complete(future_result, change_parameter_timeout_)`. However to me it's still not clear what should the filter do if a node is offline. I mean, service request fail by timeout and then? Should I throw an error? While log seems not much substantial, as I was saying this seems to be a little bit much. What if a node is not active yet or has just crashed and will be up again in a few seconds?
Anyway, it is only a matter of choices. If you want this to work only with nodes that are always up, it's ok! 
By the way I could also add another parameter to somehow distinguish between "always-up" nodes and "discontinuous" ones (for instance plugin). Or adding a "global" timeout (not for the service request timeout but for the node in general, for instance 10 seconds timeout after the first service request timeout... this would implies some re-call logic, even when there's no binary state change). Or simply logging but also exposing error information for any third part diagnostic nodes... I really don't know, just shooting ideas

Reply:
There is no need to add global timeouts and try to re-send the service again. This should be a responsibility of DDS implementation, and if some node is not online for some reasons (e.g. bad network connection), DDS to re-try to deliver the request to the given node within a timeout. So, there is no need to do the same job twice.

Regarding throwing an exception, this is a good question. Changing the state of node - is a serious situation that might break full logic working mechanisms on the end-developer side. Imagine, if you have some handling "NodeA" that is having two sub-algorithms: robot local planning and hand manipulation. By moving into some marked area, the BinaryFilter decides to stop the robot and send a binary state to "NodeA" to switch algorithm from navigation -> to hand manipulation. If this signal won't be delivered to "NodeA", it will continue to working as previously, so the normal operation logic will break.
Additionally, this situation could be treated as ROS-parameter setting error, which is also always going with exceptions throwing.

So from this, I'd more inclined that we need to throw an error in service failure situation (and thus, stop normal BF execution).

Reply:
I mean, I agree a log is not enough, while an exception may be too much in some cases. What do you think of adding a parameter for letting the user decide this one?

Each parameter can have a parameter for this: something like a boolean `critical` (default `True`).
Any suggestion on the name is appreciated! It could be reverse also, like `is_optional` (default `False`).
 Configuration would be:
 ```yaml
     ParamNamespace:
       node_name: "node"
       param_name: "param"
       default_state: True
       is_critical: False
 ```
Hence the code should look like this:
```c++
    if (!change_parameters_client->wait_for_service(
        std::chrono::milliseconds(change_parameter_timeout_)))
    {
      if (param.is_critical){
        throw std::runtime_error("BinaryFilter: Service " + 
          std::string(change_parameters_client->get_service_name()) + 
          "not available!");
      }
      RCLCPP_ERROR(
        logger_, "BinaryFilter:  service %s not available. Skipping ...",
        change_parameters_client->get_service_name());
    } else {
      RCLCPP_INFO(
        logger_, "BinaryFilter:  service %s available.",
        change_parameters_client->get_service_name());
    }
  }
  ```

We could also think of having just one global parameter for the filter, without giving the chance to specify the behavior for each parameter, but I don't see great advantages a part from having less parameters.

What do you think?

Reply:
What is the example when we have the parameter of some node not changed, but we are still normally operating? Is this worth to skip this node in the config file at all, which is not reliable and not necessary for us?
I doubt about such situation: imagine - you have some node which might be offline or online. And you can not guarantee it's parameters consistency. Is it real worth to support? In this case, I believe it is better to throw to developer an exception causing an error, allowing the developer to remove this unreliable node from config at all. Since the node is unreliable, it's binary param won't be guaranteed to be consistent, and thus I do not think, this is normal situation.

In other words, not-changing / not-setting of ROS parameters - is already a critical situation which throws and error in ROS2 API. For these cases, we could compare setting the parameter through a service request and thus it could be treated with exception. 
Adding additional parameter to choose throw exceptions or not for the unreliable nodes, in my opinion - over-complicating the structure, since we are still could not rely on this flaky node parameter, so why don't throw error instead in all times?